### PR TITLE
Introduce new CSAL ftl bdev

### PR DESCRIPTION
**Pull request by MaisenbacherD**
For this PR to be merged the PR https://github.com/openebs/spdk-rs/pull/67
for spdk-rs needs to be resolved and the submodule reference must be
updated accordingly. :)

The SPDK ftl bdev allows to create a layered device with a (fast) cache
device for buffering writes that get eventually flushed out sequentially
to a base device. SDPK ftl is also known as the Cloud Storage
Acceleration Layer (CSAL).

This kind of device potentially enables the use of emerging storage
interfaces like Zoned Namespace (ZNS) or Flexible Data Placement (FDP)
capable NVMe devices. Up to this point, those NVMe command sets are not
yet supported in SPDK upstream. However, the acceleration aspect of a
fast cache device already adds value.
With future support of new devices for SDPK ftl, Mayastor would already
be capable of utilizing those features simply by upgrading SDPK.

For now, the `ftl_mount_fs` test cases are hidden behind the `ignore`
attribute because PCIe devices are required for this test until
SPDK v24.09 is picked up, which introduces variable sector size emulation
for ftl devices.
To run these tests use the following:
```
RUST_LOG=TRACE cargo test -- --test-threads 1 --test ftl_mount_fs --nocapture --ignored
```

This patch introduces a new ftl device uri scheme:
```
ftl:///<ftl-device-name>?bbdev=<dev-uri-nested-encoding>&cbdev=<dev-uri-nested-encoding>
```

`<dev-uri-nested-encoding>` can be any already valid device uri where '?'
are replaced with '!' and '&' are replaced with '|'.
With SPDK v24.05 only PCIe devices will work where the cache device is
formatted to 4KiB+64B LBA format and the base device to 4KiB LBA format.

From SPDK v24.09 on, any device with a block size of 4KiB will work for
both cache and base device.

CSAL paper reference:
https://dl.acm.org/doi/10.1145/3627703.3629566

### CODE REVIEW

**Comment by MaisenbacherD** [comment]
Discussion question: The `spdk_ftl_conf` allows us to set configurations like `core_mask` and `l2p_dram_limit`. Are those configurations worth exposing through the ftl uri, or should we go with the default configuration?

File: doc/test.md
Comment (tiagolobocastro) [inline doc/test.md:83]: Maybe in the future we can have a feature flag for PCIe tests which can be run on a dedicated test system :)
Btw, can we add virtual PCIe devices with qemu? That could be a way of running those tests without having spare pcie nvme devices

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs]: did you run cargo fmt?

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs:79]: Add simple doc-comment explaining what `Ftl` is?

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs]: Try to start with a `C`apital and end with a `.` - makes the auto-generated rust docs look better IIRC

```suggestion
    /// The name of the ftl-bdev we created.
```

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs]: What is VSS ?
DIX is the data integrity stuff? So the cache bdev has to support blocksize + 8B, for example?
Any other requirements for this cache device? Does it need to be non-volatile so ex it can't be a ram device?

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs]: Can you document how the uri looks like? for example like for https://github.com/openebs/mayastor/blob/6374141862abf68b848e54a5dea2072d6c064819/io-engine/src/bdev/nx.rs#L1

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs]: This is because we can't have those in the "sub-uris" ? Would be good to document that at the top of the file as well.
The nx and lvs probably have the same issue, maybe I never tried using query parameters there
Can around this as well by using percent encoding?

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs]: Is this needed?

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs]: This is leaking memory?
Does it have to be mut or can we use a reference? example:
```suggestion
        let l_name =
            CString::new(self.name.clone()).unwrap();
```
```
 ftl_conf.name = l_name.as_ptr();
```

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs]: We don't have access to `SPDK_FTL_MODE_CREATE` ?

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs:288]: nit:
```suggestion
        debug!("{self:?}: deleting");
```

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs]: nit: you can transmute Option to Result or use let else to reduce indentation on the happy path

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [inline io-engine/src/bdev/ftl.rs]: Shouldn't we reverse the order here?
Destroy the backing bdevs first, and then the self bdev?

File: io-engine/src/core/env.rs
Comment (tiagolobocastro) [inline io-engine/src/core/env.rs:1135]: Pardon my ignorance, this num_entries existing already?
What is this tracing actually doing here?

File: io-engine/tests/ftl_mount_fs.rs
Comment (tiagolobocastro) [inline io-engine/tests/ftl_mount_fs.rs]:
```suggestion
//! TODO: with spdk v24.09 ftl devices support variable sector size emulation. At that point we can
//! enable testing with aio devices, which do not require a metadata section in the LBA format.
//! For now we are hiding this test case behind the `ignore` attribute.
```

Comment (tiagolobocastro) [reply io-engine/tests/ftl_mount_fs.rs]: Could you please point me to the commit which enables this emulation, just for my curiosity? :)

File: doc/test.md
Comment (MaisenbacherD) [reply doc/test.md:83]: Do you mean a rust feature flag, such that PCIe support can be selected at compile time?
It makes sense to me! I briefly thought about adding a feature flag because of the failing test case but went with the `ignore` attribute for simplicity.  
I can rework this to be a rust feature flag within this PR, however, if I attempt this (https://doc.rust-lang.org/cargo/reference/features.html) I get the following error:
```
error: failed to parse manifest at `/home/dennis/src/mayastor/Cargo.toml`

Caused by:
  this virtual manifest specifies a `features` section, which is not allowed
```
Do you know what needs to be done here?

A QEMU NVMe emulated device via PCIe would work for testing. One just needs to make sure that the cache device is configured with `ms=8` https://qemu-project.gitlab.io/qemu/system/devices/nvme.html#metadata.

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs]: Yes I did, but it did not merge the spdk_rs prefixes for some reason 🤔
I fixed it manually.

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs:79]: Done :)
Also improved the docs comment on top of the file.

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs]: I improved the docs comment on top of the file.

VSS = Variable Sector Size
DIX = Data Integrity Extension (aka Protection Information (PI))

The cache bdev needs to have an LBA format of 4KiB data size and 64B metadata size.
It also needs to be non-volatile memory if you don't want to lose data :)

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs]: Yes, it is because we can't have the `sub-uris`. I have to try the percent encoding. Lets leave this conversation as unresolved for now :)

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs]: Removed. This was a relict during development.

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs]: Good catch, yes it is leaking memory! I reworked this section.

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs]: Thanks for addressing this!
I tried importing `SPDK_FTL_MODE_CREATE` via `enum spdk_ftl_mode` which is located in `spdk/include/spdk/ftl.h` alongside `struct spdk_ftl_conf` which I am also using here.

However, when doing

```
use spdk_rs::{
    ffihelper::errno_result_from_i32,
    libspdk::{
        bdev_ftl_create_bdev, bdev_ftl_delete_bdev, ftl_bdev_info,
        spdk_ftl_conf, spdk_ftl_get_default_conf, spdk_ftl_mode,
    },
    UntypedBdev,
};
```

I am getting the following error:

```
error[E0432]: unresolved import `spdk_rs::libspdk::spdk_ftl_mode`
  --> io-engine/src/bdev/ftl.rs:44:51
   |
44 |         spdk_ftl_conf, spdk_ftl_get_default_conf, spdk_ftl_mode,
   |                                                   ^^^^^^^^^^^^^
   |                                                   |
   |                                                   no `spdk_ftl_mode` in `libspdk`
   |                                                   help: a similar name exists in the module: `spdk_ftl_conf`
```

What am I doing wrong here? I can't figure out what to change in spdk-rs' build.rs to include this enum.

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs:288]: Can't do that:
```
error[E0382]: borrow of moved value: `self`
   --> io-engine/src/bdev/ftl.rs:244:17
    |
243 |     async fn destroy(self: Box<Self>) -> Result<(), Self::Error> {
    |                      ---- value moved here
244 |         debug!("{self:?}: deleting");
    |                 ^^^^^^^^ value borrowed here after move

```

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs]: `bdev_ftl_delete_bdev` requires the backing bdev's to be present.

File: io-engine/src/core/env.rs
Comment (MaisenbacherD) [reply io-engine/src/core/env.rs:1135]: `num_entries` already exists in `MayastorEnvironment` which loosely resembles `spdk_app_opts` and indicates the number of trace entries per core.
If this `num_entries > 0` we initialize the SPDK tracing framework.

File: io-engine/tests/ftl_mount_fs.rs
Comment (MaisenbacherD) [reply io-engine/tests/ftl_mount_fs.rs]: The commit which enables this:
https://review.spdk.io/gerrit/c/spdk/spdk/+/19628/82

You can also look at the `Submitted together` commits that did the actual work :)

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs]: The percent-encoding should be possible. However, the user specified URIs will get very convoluted:
See the very simple example (without additional options in the nested URIs) with my proposed encoding vs percent-encoding:
```
ftl:///ftl-1?bbdev=pcie:///0000:01:00.0&cbdev=pcie:///0000:02:00.0
```
vs.
```
ftl:///ftl-1?bbdev=pcie%3A%2F%2F%2F0000%3A01%3A00.0&cbdev=pcie%3A%2F%2F%2F0000%3A02%3A00.0
```

What do you think how we should proceed?

File: doc/test.md
Comment (tiagolobocastro) [reply doc/test.md:83]: > Do you know what needs to be done here?

The feature flag would need to be added to the io-engine crate, see some examples: https://github.com/openebs/mayastor/blob/2825e5eb8042999718c6017229385b731f9b9b5a/io-engine/Cargo.toml#L12

I'll leave it up to you, fine to keep ignore for now as well.
<br>

> A QEMU NVMe emulated device via PCIe would work for testing. One just needs to make sure that the cache device is configured with ms=8 https://qemu-project.gitlab.io/qemu/system/devices/nvme.html#metadata.

Maybe something we may be able to do here then. We run the python BDD tests in a vm, maybe we can configure it with an emulated PCIe device and be able to run these tests then?
Alternatively we can setup VM's on demand from the tests themselves? With nix we can setup test vm's easily, example: https://github.com/openebs/mayastor-extensions/blob/develop/tests/helm/test.nix 
But I don't think there's emulated nvme support there...

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [reply io-engine/src/bdev/ftl.rs]: Can we use a mix, example, only use percent decoding for the ? and & ?

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [reply io-engine/src/bdev/ftl.rs]: Do you need to add it to `allowlist_type` on the spdk-rs build.rs?

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [reply io-engine/src/bdev/ftl.rs:288]: oh, interesting...

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [reply io-engine/src/bdev/ftl.rs]: nvm I must have been day dreaming, makes perfect sense to destroy the top-down...

**Review by tiagolobocastro — APPROVED** [review]

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs]: Thanks that works!
I was so close :D While testing I had already added the `rustified_enum` but forgot the `allowlist_type`.
I opened another PR on spdk-rs for this: https://github.com/openebs/spdk-rs/pull/69

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs]: Yes this works. I rewrote this parsing section and now we can use partial percent-encoding for `?` and `&` . Optionally you can also use percent-encoding for the entire sub-uri.
Thanks for this suggestion :)

File: doc/test.md
Comment (MaisenbacherD) [reply doc/test.md:83]: I now introduced a new feature flag :)


Yes we should definitely look into running the test in an emulated QEMU environment or something similar.

That on demand VM setup you linked looks interesting. It would take a bit time to play with it and figure out the capabilities.
Should we address the CI setup within this PR?

Comment (tiagolobocastro) [reply doc/test.md:83]: We can do in another PR, let's get this one merged :)

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [reply io-engine/src/bdev/ftl.rs]: Awesome :)

File: io-engine/src/bdev/ftl.rs
Comment (tiagolobocastro) [reply io-engine/src/bdev/ftl.rs]: Great, approved that one too!

**Comment by tiagolobocastro** [comment]
bors try

**Comment by MaisenbacherD** [comment]
Addressed the linter errors :)

File: doc/test.md
Comment (MaisenbacherD) [reply doc/test.md:83]: I will talk with you outside of this PR to discuss this :)

**Comment by dsharma-dc** [comment]
bors try

**Comment by dsharma-dc** [comment]
Linter failures still. Could you please run below script in your workspace:
`./scripts/rust-style.sh`

Please refer https://gist.github.com/dsharma-dc/a14488638a7815561571882759e7921e

**Comment by MaisenbacherD** [comment]
> Linter failures still. Could you please run below script in your workspace:

Thanks for the hint :) I just ran the script.

File: io-engine/src/bdev/ftl.rs
Comment (dsharma-dc) [inline io-engine/src/bdev/ftl.rs]:
```suggestion
//! uuid: A UUID that can be set to reference the resulting SPDK bdev.
```

**Comment by tiagolobocastro** [comment]
bors try

**Comment by MaisenbacherD** [comment]
I updated the spdk-rs submodule reference to the merged commit from https://github.com/openebs/spdk-rs/pull/69 :)

**Comment by MaisenbacherD** [comment]
bors try

**Comment by tiagolobocastro** [comment]
You should be able to use bors with this:
bors d+

**Comment by MaisenbacherD** [comment]
Resolved some more linter issues :)

**Comment by MaisenbacherD** [comment]
bors d+

**Comment by MaisenbacherD** [comment]
bors try

File: io-engine/src/bdev/ftl.rs
Comment (dsharma-dc) [inline io-engine/src/bdev/ftl.rs:259]: Shouldn't we destroy the base bdev and cache bdev created earlier here, in case of ftl bdev creation failed?

Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs:259]: I agree :) Will address this tomorrow.

**Comment by MaisenbacherD** [comment]
Updated spdk-rs submodule reference https://github.com/openebs/spdk-rs/pull/70#pullrequestreview-2465784790

File: io-engine/src/bdev/ftl.rs
Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs:259]: @dsharma-dc and @tiagolobocastro
Would it make sense to implement the drop trait for `Ftl` here where we always `bdev_destroy` `bbdev` and `cbdev`?
Further up in the code the creation of `cbdev` might fail and then we would need to destroy `bbdev` as well.

Comment (tiagolobocastro) [reply io-engine/src/bdev/ftl.rs:259]: Possibly though may be difficult as `Drop` is blocking. We could potentially just schedule the deletion to happen asynchronously, though this may become an awkward api.
Another question here is, what if the b and c devices are exist?  Presumably this fails on `bdev_create`?

Comment (MaisenbacherD) [reply io-engine/src/bdev/ftl.rs:259]: > Possibly though may be difficult as Drop is blocking. We could potentially just schedule the deletion to happen asynchronously, though this may become an awkward api.

Okay, then I will handle those cases individually :)

> Another question here is, what if the b and c devices are exist? Presumably this fails on bdev_create?

Yes it is going to fail, and I would propose to keep it like this because we don't want to accidentally reformat a drive that is already in use :)

**Comment by MaisenbacherD** [comment]
@dsharma-dc Thanks for your comments :)
I addressed them.

**Comment by MaisenbacherD** [comment]
bors try

**Review by dsharma-dc — APPROVED** [review]

**Comment by MaisenbacherD** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
bors merge

**Comment by tiagolobocastro** [comment]
doh, had missed a couple
bors merge

**Pull request merged**
